exchanges/exchanges.go

```diff
@@ -1821,9 +1956,367 @@ func (r PoloniexCandlestickResponse) translate(bin candlestickKey) Candlesticks
+// are of the "o" type.
+func (poloniex *PoloniexExchange) processTrade(tradeParams []interface{}) (*poloniexOrder, int, error) {
+	if len(tradeParams) != 6 {
+		return nil, -1, fmt.Errorf("Not enough parameters in poloniex trade notification. given: %d", len(tradeParams))
+	}
+	floatDir, ok := tradeParams[2].(float64)
+	if !ok {
+		return nil, -1, fmt.Errorf("failed to type convert poloniex orderbook update direction")
+	}
+	direction := (int(floatDir) + 1) % 2
+	priceStr, ok := tradeParams[3].(string)
+	if !ok {
+		return nil, -1, fmt.Errorf("failed to type convert poloniex orderbook update price")
+	}
+	price, err := strconv.ParseFloat(priceStr, 64)
+	if err != nil {
+		return nil, -1, fmt.Errorf("failed to convert poloniex orderbook update price to float: %v", err)
+	}
+	volStr, ok := tradeParams[4].(string)
+	if !ok {
+		return nil, -1, fmt.Errorf("failed to type convert poloniex orderbook update volume")
+	}
+	volume, err := strconv.ParseFloat(volStr, 64)
+	if err != nil {
+		return nil, -1, fmt.Errorf("failed to convert poloniex orderbook update volume to float: %v", err)
+	}
+	trade := &poloniexOrder{
+		price:  price,
+		volume: volume,
+	}
+	return trade, direction, nil
+}
+
+// Poloniex's WebsocketProcessor. Handles messages of type "i", "o", and "t".
+func (poloniex *PoloniexExchange) processWsMessage(raw []byte) {
+	msg := make([]interface{}, 0)
+	err := json.Unmarshal(raw, &msg)
+	if err != nil {
+		poloniex.setWsFail(err)
+		return
+	}
+	if len(msg) == 1 {
+		// Likely a heatbeat
+		code, ok := msg[0].(float64)
+		if !ok {
+			poloniex.setWsFail(fmt.Errorf("non-integer single-element poloniex response of implicit type %T", msg[0]))
+			return
+		}
+		intCode := int(code)
+		if intCode == poloniexHeartbeatCode {
+			return
+		}
+		poloniex.setWsFail(fmt.Errorf("unknown code in single-element poloniex response: %d", intCode))
+		return
+	}
+	if len(msg) < 3 {
+		poloniex.setWsFail(fmt.Errorf("poloniex websocket message had unexpected length %d", len(msg)))
+		return
+	}
+
+	responseList, ok := msg[2].([]interface{})
+	if !ok {
+		poloniex.setWsFail(fmt.Errorf("poloniex websocket message type assertion failure: %T", msg[2]))
+		return
+	}
+
+	code := firstCode(responseList)
+
+	if code == poloniexInitialOrderbookKey {
+		poloniex.processWsOrderbook(responseList)
+		state := poloniex.state()
+		if state != nil { // Only send update if price has been fetched
+			depth := poloniex.wsDepths()
+			poloniex.Update(&ExchangeState{
+				Price:        state.Price,
+				BaseVolume:   state.BaseVolume,
+				Volume:       state.Volume,
+				Change:       state.Change,
+				Depth:        depth,
+				Candlesticks: state.Candlesticks,
+			})
+		}
+		return
+	}
+
+	if code != poloniexOrderUpdateKey && code != poloniexTradeUpdateKey {
+		poloniex.setWsFail(fmt.Errorf("Unexpected code in first element of poloniex websocket response list: %s", code))
+		return
+	}
+
+	poloniex.orderMtx.RLock()
```

Comment:
What is being protected against concurrent writes in this section?

Reply:
Thanks. When I was processing trades as updates to the order book (incorrectly), the lock was needed, but I corrected that. Dropping the lock now.